pkg/scheduler/core/spreadconstraint/select_groups.go

```diff
@@ -0,0 +1,202 @@
+package spreadconstraint
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+// GroupInfo indicate the group info.
+type GroupInfo struct {
+	name   string
+	value  int
+	weight int64
+}
+
+type dfsPath struct {
+	id     int
+	groups []*GroupInfo
+	weight int64
+	value  int
+}
+
+func (path *dfsPath) next() *dfsPath {
+	path.id++
+	result := new(dfsPath)
+	*result = *path
+	if path.groups != nil {
+		result.groups = make([]*GroupInfo, len(path.groups))
+		copy(result.groups, path.groups)
+	}
+	for _, group := range result.groups {
+		result.weight += group.weight
+		result.value += group.value
+	}
+	result.sortGroups()
+	return result
+}
+
+func (path *dfsPath) enqueue(group *GroupInfo) {
+	path.groups = append(path.groups, group)
+}
+
+func (path *dfsPath) popLast() {
+	path.groups = path.groups[:path.length()-1]
+}
+
+func (path *dfsPath) length() int {
+	return len(path.groups)
+}
+
+func (path *dfsPath) sortGroups() {
+	sort.Slice(path.groups, func(i, j int) bool {
+		if path.groups[i].weight != path.groups[j].weight {
+			return path.groups[i].weight > path.groups[j].weight
+		}
+
+		return path.groups[i].name < path.groups[j].name
+	})
+}
+
+func (path *dfsPath) matchSubPath(subPath *dfsPath) bool {
+	if subPath.length() >= path.length() {
+		return false
+	}
+
+	for i, group := range subPath.groups {
+		if path.groups[i].name != group.name {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (path *dfsPath) String() string {
+	var groupNames []string
+	for _, group := range path.groups {
+		groupNames = append(groupNames, group.name)
+	}
+	return fmt.Sprintf("[%s]", strings.Join(groupNames, "->"))
+}
+
+// selectGroups select best groups in all groups.
+func selectGroups(groups []*GroupInfo, minConstraint, maxConstraint, target int) []*GroupInfo {
```

Comment:
The signature `target` might be not so clear enough from the code readability point of view.